executor/read/pattern_executor.rs

```diff
@@ -0,0 +1,188 @@
+/*
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at https://mozilla.org/MPL/2.0/.
+ */
+
+use lending_iterator::LendingIterator;
+use storage::snapshot::ReadableSnapshot;
+
+use crate::{
+    batch::{FixedBatch, FixedBatchRowIterator},
+    error::ReadExecutionError,
+    pipeline::stage::ExecutionContext,
+    read::{
+        collecting_stage_executor::CollectedStageIterator, nested_pattern_executor::SubQueryResult,
+        step_executor::StepExecutors,
+    },
+    ExecutionInterrupt,
+};
+
+#[derive(Clone)]
+pub(super) struct BranchIndex(pub usize);
+
+#[derive(Clone)]
+pub(super) struct InstructionIndex(pub usize);
+
+impl InstructionIndex {
+    fn next(&self) -> InstructionIndex {
+        InstructionIndex(self.0 + 1)
+    }
+}
+
+pub(super) enum ControlInstruction {
+    Start(FixedBatch),
+
+    ExecuteImmediate(InstructionIndex),
+
+    MapRowBatchToRowForSubQuery(InstructionIndex, FixedBatchRowIterator),
+    ExecuteSubQuery(InstructionIndex, BranchIndex),
+
+    CollectingStage(InstructionIndex),
+    StreamCollected(InstructionIndex, CollectedStageIterator),
```

Comment:
I think naming the second to imply the 'Done'ness of the collection a bit more precisely could help? `CollectingStageOutput` (tie it directly to `CollectingStage`)